### PR TITLE
Removed `ShutdownMode`. Now always behaves like original `Reload`

### PR DESCRIPTION
**Pull request by lostmsu**
### What does this implement/fix? Explain your changes.

This removes all shutdown modes, except `Reload` (which you don't need to specify, because it is the only one left).

This means Python C runtime is never actually shut down when .NET calls `PythonEngine.Shutdown()`. Instead, anything from .NET exposed to Python before `Shutdown` becomes unavailable until `PythonEngine.Initialize()` is called again (which can be done from a different `AppDomain`).

### Any other comments?

Also in this change:
- dropped Python 3.6 support
- fixed Python derived types not being decrefed when an instance is deallocated
- reduced time and amount of storage needed for runtime reload
- removed circular reference loop between Type <-> `ConstructorBinding`(s)
+ exposed `Runtime.TryCollectingGarbage`

A review from @amos402 would be welcome

### CODE REVIEW

File: .github/workflows/main.yml
Comment (filmor) [inline .github/workflows/main.yml:15]: Is there any particular reason to do this now?

Comment (lostmsu) [reply .github/workflows/main.yml:15]: It [EOLs in 3 days](https://endoflife.date/python). Does not have to be a part of this PR, just thought I'd clean up the matrix. Considering EOL, IMHO we should stop supporting it in 3.0.

**Comment by lostmsu** [comment]
@filmor looks like I got tests to pass, so this is ready.

File: .github/workflows/main.yml
Comment (filmor) [reply .github/workflows/main.yml:15]: I don't see a compelling reason to remove it now, though. If we see it acting up, we can remove support, but that means (to me) that we also adjust the package metadata and remove the interop file. Please revert this.

Comment (lostmsu) [reply .github/workflows/main.yml]: @filmor OK, moved into https://github.com/pythonnet/pythonnet/issues/1640

File: src/runtime/pythonengine.cs
Comment (filmor) [inline src/runtime/pythonengine.cs:317]: 1. The error should be something like `"Python.NET runtime is already initialized"`
2. We should probably use `PyEval_GetBuiltins` everywhere

File: src/runtime/pytype.cs
Comment (filmor) [inline src/runtime/pytype.cs:114]: This is not right. It's documented to be an `unsigned long`: https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_flags

File: src/runtime/runtime.cs
Comment (filmor) [inline src/runtime/runtime.cs:360]: This is my only real gripe with this: Embedders would not be able to finalise the Python runtime at all anymore. Maybe we should provide this (`Shutdown()` + `Py_Finalize()`) as a possible footgun with a long name?

File: src/runtime/runtime.cs
Comment (filmor) [inline src/runtime/runtime.cs:1926]: The minimal version we support right now is 3.6.

File: src/runtime/runtime.cs
Comment (lostmsu) [reply src/runtime/runtime.cs:360]: We can always add it if there is a strong request. Instead what happened is people opening bugs about crashes after restarting runtime because NumPy does not support it. I'd rather avoid it.

Moreover, the behavior of the whole thing when either of runtimes is shut down completely is pretty much undefined.

File: src/runtime/runtime.cs
Comment (lostmsu) [reply src/runtime/runtime.cs]: Reverted constant name, edited constant text instead

File: src/runtime/pythonengine.cs
Comment (lostmsu) [reply src/runtime/pythonengine.cs]: Changed error message, switched to `PyEval_GetBuiltins` here

**Comment by lostmsu** [comment]
@filmor I am waiting for an explicit approval or more comments

**Review by filmor — APPROVED** [review]

**Pull request merged**
